File: payroll-api/src/main/java/com/company/payroll/util/JwtTokenUtils.java
package com.company.payroll.util;

import java.security.Key;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Base64;
import java.util.Date;
import java.util.Map;

import javax.crypto.spec.SecretKeySpec;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.company.payroll.model.JwtConfig;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;

@Component
public class JwtTokenUtils {	
	private SignatureAlgorithm ALG = SignatureAlgorithm.HS256;
	private Instant issuedAt = Instant.now().truncatedTo(ChronoUnit.SECONDS);

	@Autowired
	private JwtConfig jwtTokenProperties;

	/**
	 * Generate token
	 * @param Claims
	 * @return Token
	 */
	public String generateToken(String username, Map<String, Object> claims) {
		Key key = new SecretKeySpec(Base64.getEncoder().encode(jwtTokenProperties.getKey().getBytes()), 
									ALG.getJcaName());
		Instant expiration = issuedAt.plus(jwtTokenProperties.getTtl(), 
											ChronoUnit.SECONDS);

		return Jwts.builder()
					.setClaims(claims)
					.setSubject(username)
					.setExpiration(Date.from(expiration))
					.signWith(key, ALG)
					.compact();
	}
	
	/**
	 * Get claims from token
	 * @param token
	 * @return Claims
	 */
	public Claims getClaims(String token) {
		return Jwts.parserBuilder()
				.setSigningKey(Base64.getEncoder().encode(jwtTokenProperties.getKey().getBytes()))
				.build()
				.parseClaimsJws(token)
				.getBody();
	}
	
	/**
	 * Get subject username
	 * @param token
	 * @return
	 */
	public String getUsername(String token) {
		Claims claims = getClaims(token);
		
		return claims.getSubject();
	}
	
	/**
	 * validate token
	 * @param token
	 * @return
	 */
	public boolean validateToken(String token) {
		if(getUsername(token) != null && isExpired(token)) {
			return true;
		}
		
		return false;
	}
	
	/**
	 * Validate token is expired
	 * @param token
	 * @return
	 */
	public boolean isExpired(String token) {
		Claims claims = getClaims(token);
		return claims.getExpiration().after(Date.from(Instant.now()
															 .truncatedTo(ChronoUnit.SECONDS)));
	}
}

File: kotlin-payroll/src/main/kotlin/com/company/payroll/util/JwtTokenUtils.kt
package com.company.payroll.util

import com.company.payroll.model.JwtTokenProperties
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.security.Key
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.*
import javax.crypto.spec.SecretKeySpec

@Component
class JwtTokenUtils(@Autowired private val jwtTokenProperties: JwtTokenProperties) {
  private val alg: SignatureAlgorithm = SignatureAlgorithm.HS256
  private val issuedAt = Instant.now().truncatedTo(ChronoUnit.SECONDS)

  /**
   * Generate token
   * @param claims
   * @return Token
   */
  fun generateToken(username: String, claims: Map<String, Any>): String {
    val key: Key = SecretKeySpec(Base64.getEncoder().encode(jwtTokenProperties.key.toByteArray()),
        alg.jcaName)
    val expiration = issuedAt.plus(jwtTokenProperties.ttl,
        ChronoUnit.SECONDS)
    return Jwts.builder()
        .setClaims(claims)
        .setSubject(username)
        .setExpiration(Date.from(expiration))
        .signWith(key, alg)
        .compact()
  }

  /**
   * Get claims from token
   * @param token
   * @return Claims
   */
  fun getClaims(token: String): Claims {
    return Jwts.parserBuilder()
        .setSigningKey(Base64.getEncoder().encode(jwtTokenProperties.key.toByteArray()))
        .build()
        .parseClaimsJws(token)
        .body
  }

  /**
   * Get subject username
   * @param token
   * @return
   */
  fun getUsername(token: String): String? {
    val claims = getClaims(token)
    return claims.subject
  }

  /**
   * validate token
   * @param token
   * @return
   */
  fun validateToken(token: String): Boolean {
    return getUsername(token) != null && isExpired(token)
  }

  /**
   * Validate token is expired
   * @param token
   * @return
   */
  fun isExpired(token: String): Boolean {
    val claims = getClaims(token)
    return claims.expiration.after(Date.from(Instant.now()
        .truncatedTo(ChronoUnit.SECONDS)))
  }
}